crates/oxc_linter/src/rules/eslint/no_redeclare.rs

```diff
@@ -0,0 +1,159 @@
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::{self, Error},
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_semantic::AstNode;
+use oxc_span::{Atom, Span};
+
+use crate::{context::LintContext, globals::BUILTINS, rule::Rule};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-redeclare): '{0}' is already defined.")]
+#[diagnostic(severity(warning))]
+struct NoRedeclareDiagnostic(
+    Atom,
+    #[label("'{0}' is already defined.")] pub Span,
+    #[label("It can not be redeclare here.")] pub Span,
+);
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-redeclare): '{0}' is already defined as a built-in global variable.")]
+#[diagnostic(severity(warning))]
+struct NoRedeclareAsBuiltiInDiagnostic(
+    Atom,
+    #[label("'{0}' is already defined as a built-in global variable.")] pub Span,
+);
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-redeclare): '{0}' is already defined by a variable declaration.")]
+#[diagnostic(severity(warning))]
+struct NoRedeclareBySyntaxDiagnostic(
+    Atom,
+    #[label("'{0}' is already defined by a variable declaration.")] pub Span,
+    #[label("It can not be redeclare here.")] pub Span,
+);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoRedeclare {
+    built_in_globals: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow variable redeclaration
+    ///
+    /// ### Why is this bad?
+    ///
+    /// n JavaScript, it’s possible to redeclare the same variable name using var. This can lead to confusion as to where the variable is actually declared and initialized.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// var a = 3;
+    /// var a = 10;
+    /// ```
+    NoRedeclare,
+    correctness
+);
+
+impl Rule for NoRedeclare {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let built_in_globals = value
+            .get(0)
+            .and_then(|config| config.get("builtinGlobals"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or(false);
+
+        Self { built_in_globals }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
```

Comment:
The perf regression indicates that running this on every binding identifier is costly.

We may be able to use `run_once` and print all the diagnostics from interpreting `redeclare_variables`.